include/fmt/core.h

```diff
@@ -2204,20 +2204,32 @@ constexpr auto to_ascii(Char c) -> char {
   return c <= 0xff ? static_cast<char>(c) : '\0';
 }
 
+// Returns the length of a codepoint. Returns 0 for invalid codepoints.
 FMT_CONSTEXPR inline auto code_point_length_impl(char c) -> int {
   return "\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\0\0\0\0\0\0\0\0\2\2\2\2\3\3\4"
       [static_cast<unsigned char>(c) >> 3];
 }
+// Returns the length of a codepoint. Returns 1 for invalid codepoints.
+// This is equivalent to
+//
+// int len = code_point_length_impl(c);
+// return len + !len;
+//
+// This is useful because it allows the compiler to check that the
+// length is within the range [1, 4]
+FMT_CONSTEXPR inline auto code_point_length_impl_2(char c) -> int {
+  return static_cast<int>((0x3a55000000000000ull >> (2 * (static_cast<unsigned char>(c) >> 3))) & 0x3) + 1;
```

Comment:
I think this logic deserves some explanation.

Reply:
Hi! Sorry for the delay.

code_point_length_impl is initially implemented like this, which gets the length of a code point based on the first character in that codepoint.

```cpp
MT_CONSTEXPR inline auto code_point_length_impl(char c) -> int {
  return "\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\0\0\0\0\0\0\0\0\2\2\2\2\3\3\4"
      [static_cast<unsigned char>(c) >> 3];
}
```

For valid codepoints, we return the length of the codepoint, and for invalid codepoints we return a length of 0.

When used in `code_point_length`, the length of invalid codepoints gets converted to 1, so we don't actually need to distinguish between valid and invalid codepoints for `code_point_length`.

As a result, we could write it like this:

```cpp
MT_CONSTEXPR inline auto code_point_length_impl_2(char c) -> int {
  return "\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\2\2\2\2\3\3\4"
      [static_cast<unsigned char>(c) >> 3];
}
```

Now we have 4 possible return values: [1, 2, 3, 4]. If we instead use [0, 1, 2, 3] and just add one, it looks like this now:

```cpp
MT_CONSTEXPR inline auto code_point_length_impl_2(char c) -> int {
  return ("\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\1\1\1\1\2\2\3"
      [static_cast<unsigned char>(c) >> 3]) + 1;
}
```

Because this array only contains values 0..3, each value can be represented in 2 bits. This means that the entire array fits in a 64 bit integer.

We can get that integer like this:

```py
arr = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,2,2,3]
result = 0
pow = 1

for x in arr:
    result += pow * x
    pow *= 4

print(f'0x{result:x}')
```
This gives us `0x3a55000000000000`.

That logic basically indexes into this integer, treating it as an array of 2-bit values, resulting in the final implementation:

```cpp
FMT_CONSTEXPR inline auto code_point_length_impl_2(char c) -> int {
  return static_cast<int>((0x3a55000000000000ull >> (2 * (static_cast<unsigned char>(c) >> 3))) & 0x3) + 1;
}
```

This has two benefits:

- The compiler recognizes that the return value is always between 1 and 4, so it gets rid of the stringop-overflow warning
- We no longer need to access memory when computing the length of a codepoint

That being said the code could definitely be less ugly.

I'll apply your recommendations with regard to merging `code_point_length_impl` into utf8_decode, and merging `code_point_length_impl_2` into `code_point_length`. I haven't noticed any similar false positives in utf8_decode.

Reply:
Makes sense, thanks for the explanation.